torch/_inductor/ir.py

```diff
@@ -4231,34 +4231,24 @@ def get_alias_names(self):
         return [self.inputs[0].codegen_reference()]
 
 
-class AllReduce(ExternKernel):
-    def __init__(
-        self,
-        layout,
-        inputs,
-        constant_args=(),
-    ):
+class CollectiveKernel(ExternKernel):
+    """
+    Each CollectiveKernel should follow the patterns
+    - it writes into a given output buffer
+    - the kernel delegates into c10d processgroup, which returns a 'work' obj
+    - the work obj is registered via _register_tensor_work so it can be waited on later
+    """
+
+    def __init__(self, layout, inputs, constant_args):
         super().__init__(None, layout, inputs, constant_args)
         self.name = V.graph.register_buffer(self)
 
     def should_allocate(self):
         return True
 
-    @classmethod
-    def create(
-        cls, x: "TensorBox", reduce_op: str, tag: str, ranks: List[int], group_size: int
-    ):
-        x = cls.realize_input(x)
-
-        # is there a difference between literally using x.data.layout below, vs
-        # creating a new one that has the same properties?
-        new_layout = FlexibleLayout(x.get_device(), x.get_dtype(), x.get_size())
-
-        return AllReduce(
-            layout=new_layout,
-            inputs=[x],
-            constant_args=[reduce_op, tag, ranks, group_size],
-        )
+    def codegen_collective(self, wrapper, output_name, input_names):
+        # factor so the boilerplate can be handled in CollectiveKernel.codegen
+        raise NotImplementedError("Must implement")
 
     def codegen(self, wrapper):
         wrapper.add_import_once("import torch.distributed as dist")
```

Comment:
is it ok to always import `import _str_to_reduce_op` as some op might not need it?

Reply:
yea, its always ok, i thought about optimizing it but its hard to imagine it would actually matter.  (Since it is already limited to 'once per program')

otoh, there is some more serious concern about having `_find_pg_...` called multiple times (once per op), which i think we should optimize but later